src/measure_theory/interval_integral.lean

```diff
@@ -1360,4 +1374,53 @@ theorem integral_deriv_eq_sub (hderiv : ∀ x ∈ interval a b, differentiable_a
   ∫ y in a..b, deriv f y = f b - f a :=
 integral_eq_sub_of_has_deriv_at (λ x hx, (hderiv x hx).has_deriv_at) hcont'
 
+/-!
+### Integration by parts
+-/
+
+lemma integral_deriv_mul_eq_sub' {u v u' v' : ℝ → ℝ} (hab : a ≤ b)
+  (hu : ∀ x ∈ Icc a b, has_deriv_at u (u' x) x)
+  (hv : ∀ x ∈ Icc a b, has_deriv_at v (v' x) x)
+  (hcu' : continuous_on u' (Icc a b)) (hcv' : continuous_on v' (Icc a b)) :
+  ∫ x in a..b, u' x * v x + u x * v' x = u b * v b - u a * v a :=
+begin
+  rw integral_eq_sub_of_has_deriv_at,
+  intros x hx;
+  simp only [interval_of_le hab, min_eq_left, max_eq_right] at hx,
+  { exact (hu x hx).mul (hv x hx) },
+  { simp only [interval_of_le hab, min_eq_left,max_eq_right],
+    exact (hcu'.mul (has_deriv_at.continuous_on (hv x hx))).add
+    ((has_deriv_at.continuous_on (hu x hx)).mul hcv') }
+end
+
+lemma integral_deriv_mul_eq_sub {u v u' v' : ℝ → ℝ}
+  (hu : ∀ x ∈ interval a b, has_deriv_at u (u' x) x)
+  (hv : ∀ x ∈ interval a b, has_deriv_at v (v' x) x)
+  (hcu : continuous_on u' (interval a b)) (hcv : continuous_on v' (interval a b)) :
+  ∫ x in a..b, u' x * v x + u x * v' x = u b * v b - u a * v a :=
+begin
+  cases le_total a b with hab hab,
+  { simp only [interval_of_le hab] at hu hv hcu hcv ⊢,
+    exact integral_deriv_mul_eq_sub' hab hu hv hcu hcv },
+  { simp only [interval_of_ge hab] at hu hv hcu hcv ⊢,
+    rw [integral_symm,integral_deriv_mul_eq_sub' hab hu hv hcu hcv, neg_sub] }
+end
+
+theorem integral_mul_deriv_eq_deriv_mul {u v u' v' : ℝ → ℝ}
+  (hu : ∀ x ∈ interval a b, has_deriv_at u (u' x) x)
+  (hv : ∀ x ∈ interval a b, has_deriv_at v (v' x) x)
+  (hcu : continuous_on u' (interval a b)) (hcv : continuous_on v' (interval a b)) :
+  ∫ x in a..b, u x * v' x = u b * v b - u a * v a - ∫ x in a..b, v x * u' x :=
+begin
+  rw [← integral_deriv_mul_eq_sub hu hv hcu hcv, ← integral_sub],
+  { apply integral_congr,
+    intros x hx,
+    simp only [mul_comm, add_comm, add_sub_assoc, sub_self, add_zero] },
+  exacts [continuous_on.interval_integrable
+      (λ x hx, ((hcu x hx).mul (hv x hx).continuous_at.continuous_within_at).add
+      ((hu x hx).continuous_at.continuous_within_at.mul (hcv x hx))),
+      continuous_on.interval_integrable
+      (λ x hx, (hv x hx).continuous_at.continuous_within_at.mul (hcu x hx))],
```

Comment:
You can get dot notation working here too.  Also, since some of these checks are performed twice, I'd be tempted to separate them out: write
```
  have hcu' : continuous_on u _ :=  λ x hx, (hu x hx).continuous_at.continuous_within_at,
  have hcv' : continuous_on v _ := λ x hx, (hv x hx).continuous_at.continuous_within_at,
```
at the beginning of the proof and then you can do
```suggestion
  { exact ((hcu.mul hcv').add (hcu'.mul hcv)).interval_integrable },
  { exact (hcv'.mul hcu).interval_integrable },
```
here.

Reply:
Or you could skip the preliminaries and use your new lemma here!
```suggestion
  { exact ((hcu.mul hv.continuous_on).add (hu.continuous_on.mul hcv)).interval_integrable },
  { exact (hv.continuous_on.mul hcu).interval_integrable },
```

Reply:
I think you missed this comment?  The point is that (1) `continuous_on.interval_integrable` accepts dot notation as well; (2) you don't need any of the `λ x hx`.

Reply:
Oops! I need to study this solution, I thought I always had to introduce variables because of the type of statement, this is so useful!

Reply:
Yes, the library has all sorts of lemmas like `continuous_on.mul`, so the variables don't need to be introduced!